File: src/java/com/distributedlife/mahjong/reference/permute/StandardSequencePermutator.java
package com.distributedlife.mahjong.reference.permute;

import com.distributedlife.mahjong.reference.data.TileSet;
import com.distributedlife.mahjong.reference.hand.HandCandidate;

import java.util.HashSet;
import java.util.Set;

public class StandardSequencePermutator extends Permutator {
    private final int from;
    private final int to;
    private String defaultSuit;

    public StandardSequencePermutator(int from, int to, String defaultSuit) {
        this.from = from;
        this.to = to;
        this.defaultSuit = defaultSuit;
    }

    @Override
    public Set<HandCandidate> permute(Set<HandCandidate> candidates) {
        return permute(candidates, defaultSuit);
    }

    @Override
    public Set<HandCandidate> permute(Set<HandCandidate> candidates, String suit) {
        Set<HandCandidate> handCandidates = new HashSet<HandCandidate>();

        for(HandCandidate candidate : candidates) {
            for (int i = from; i <= to; i++) {
                String tile = TileSet.convertTileToAppropriateSuit(suit, Integer.toString(i), candidate);

                if (candidate.canAdd(tile)) {
                    candidate.add(tile);
                }
            }

            handCandidates.add(candidate);
        }

        return handCandidates;
    }
}

File: lib/permute/standard-sequence-permutator.js
/**
 * StandardSequencePermutator - Handles permutations for runs/chows (sequences of tiles)
 */
import { Permutator } from './permutator.js';
import { convertTileToAppropriateSuit } from '../data/tile-set.js';

export class StandardSequencePermutator extends Permutator {
  constructor(from, to, defaultSuit) {
    super();
    this.from = from;
    this.to = to;
    this.defaultSuit = defaultSuit;
  }

  permute(candidates, suit = null) {
    const useSuit = suit || this.defaultSuit;
    const handCandidates = new Set();

    for (const candidate of candidates) {
      const variation = candidate.fork();
      
      for (let i = this.from; i <= this.to; i++) {
        const tile = convertTileToAppropriateSuit(useSuit, i.toString(), variation);

        if (variation.canAdd(tile)) {
          variation.add(tile);
        }
      }

      handCandidates.add(variation);
    }

    return handCandidates;
  }
}
